src/modules/scraper/service.ts

```diff
@@ -28,138 +28,104 @@ export class ScraperService {
+          configStartBlockNumber,
+          this.merkleDistributorProcessedBlockRepository,
+          true,
+        );
+
+        if (!!range) {
+          const queueMsg = { chainId, ...range };
+          await this.scraperQueuesService.publishMessage<MerkleDistributorBlocksEventsQueueMessage>(
+            ScraperQueue.MerkleDistributorBlocksEvents,
+            queueMsg,
+          );
+        }
+      } catch (error) {
+        this.logger.error(error);
+      }
+      await wait(interval);
     }
-
-    await this.scraperQueuesService.publishMessage<MerkleDistributorBlocksEventsQueueMessage>(
-      ScraperQueue.MerkleDistributorBlocksEvents,
-      {
-        chainId,
-        ...blockRange,
-      },
-    );
   }
 
-  /**
-   * Fetch the latest block numbers from all supported chains
-   */
-  public getLatestBlocks = async () => {
-    const chainIds = Object.keys(this.providers.getProviders()).map((chainId) => parseInt(chainId));
-    const blocks = await Promise.all(chainIds.map((chainId) => this.providers.getProvider(chainId).getBlock("latest")));
-    const blockNumbers = blocks.reduce(
-      (acc, block, idx) => ({ ...acc, [chainIds[idx]]: block.number }),
-      {} as Record<string, number>,
-    );
-
-    return blockNumbers;
-  };
-
   /**
    * Compute the start and the end of the next batch of blocks that needs to be processed.
    * `from` is computed depending on the latest block saved in DB || start block number defined in config file || 1
    * `to` is a block number up to the latest block number from chain, but capped at a max value. This way we avoid
-   *     huge block ranges to be processed.
+   * huge block ranges to be processed.
+   * @returns the block range or undefined if from > to
    */
-  public async determineBlockRanges(
-    latestBlocks: Record<string, number>,
-    startBlockNumbers: Record<string, { startBlockNumber: number }>,
-    blockRepository: Repository<ProcessedBlock | MerkleDistributorProcessedBlock>,
-  ) {
-    const blockRanges: Record<string, { from: number; to: number }> = {};
-
-    for (const chainIdStr of Object.keys(latestBlocks)) {
-      const chainId = parseInt(chainIdStr);
-      const configStartBlockNumber = startBlockNumbers[chainId]?.startBlockNumber;
-
-      if (!configStartBlockNumber) {
-        continue;
-      }
-
-      const blockRange = await this.determineBlockRange(
-        chainId,
-        latestBlocks[chainId],
-        configStartBlockNumber,
-        blockRepository,
-      );
-
-      if (blockRange) {
-        blockRanges[chainId] = blockRange;
-      }
-    }
-
-    return blockRanges;
-  }
-
   public async determineBlockRange(
     chainId: number,
     latestBlockNumber: number,
-    configStartBlockNumber: number,
+    startBlockNumber: number,
     blockRepository: Repository<ProcessedBlock | MerkleDistributorProcessedBlock>,
+    keepDistanceFromHead = false,
   ) {
     let previousProcessedBlock = await blockRepository.findOne({
```

Comment:
I know this isn't changed, but how do you know the previous iteration has finished being processed before the next one starts? Is there any synchronization to ensure that you don't rerun the same range a second time because the prev one hasn't finished?

Reply:
> how do you know the previous iteration has finished being processed before the next one starts

You don't know. This particular logic just defines the block ranges (`from`: last block registered in the DB; `to`: latest block number from chain) and submits it to the `BlocksEventsQueue`. The consumer attached to this queue will extract the events from that particular block range and the order is not mandatory to be maintained. The block ranges can be processed in any order